src/test/groovy/com/example/demo/api/ProductControllerTest.groovy

```diff
@@ -0,0 +1,86 @@
+package com.example.demo.api
+
+import com.example.demo.domain.*
+import com.fasterxml.jackson.core.JsonProcessingException
+import com.fasterxml.jackson.databind.ObjectMapper
+import groovy.json.JsonGenerator
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.boot.test.context.SpringBootTest
+import org.springframework.boot.test.web.client.TestRestTemplate
+import org.springframework.boot.web.server.LocalServerPort
+import org.springframework.http.HttpEntity
+import org.springframework.http.HttpHeaders
+import org.springframework.http.ResponseEntity
+import spock.lang.Specification
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class ProductControllerTest extends Specification {
+
+    @Autowired
+    ProductFacade productFacade
+
+    @Autowired
+    TestRestTemplate httpClient
+
+    @LocalServerPort
+    int port
+
+    @Autowired
+    ObjectMapper objectMapper
+
+    PriceDto dummyPrice = new PriceDto("99.99", "PLN")
+    ImageDto dummyImg = new ImageDto("https://via.placeholder.com/150")
+    DescriptionDto dummyDescr = new DescriptionDto("long description")
+    Set<TagDto> dummyTags = Set.of(new TagDto("tag1"))
+
+
+    def "should create product"() {
+        given: "Valid product request DTO"
+        def requestDto = new ProductRequestDto("testname", dummyPrice, dummyImg, dummyDescr, dummyTags)
+//        def expectedResponseDto = new ProductResponseDto("id", "testname", new PriceDto("1123123", "USD"), dummyImg, dummyDescr, dummyTags)
+        def expectedResponseDto = new ProductResponseDto("id", "testname", dummyPrice, dummyImg, dummyDescr, dummyTags)
+
+        def jsonGenerator = new JsonGenerator.Options().excludeFieldsByName("id").build()
```

Comment:
A czemu nie skorzystać z objectmappera, którego wstrzykujesz wyżej?

Reply:
Dzięki JsonGenerator można łatwo wykluczyć id z przychodzącego JSONa produktu. Po wykluczeniu id wychodzący i przychodzący JSON są takie same i wystarczy porównać stringi, bez konstruowania obiektów.
Z ObjectMapperem też można tak zrobić, ale nie jest to tak proste jak z JsonGeneratorem.